File: server/binary_protocol.py
#!/usr/bin/env python3
"""
CLAUDEPAD - Binary protocol encoder/decoder (v3.3).
Menggantikan JSON untuk command yang sering dikirim (move, scroll, click, dll).
Paket 4-10x lebih kecil dari JSON, mengurangi latency & CPU di kedua sisi.

Wire format:
  [CMD_ID: 1 byte][PAYLOAD: 0..N bytes]

Semua integer multi-byte memakai little-endian.
"""

import struct

# ── Command IDs ──────────────────────────────────────────────────────────────
CMD_MOVE       = 0x01   # dx: i16, dy: i16
CMD_CLICK      = 0x02   # button_id: u8, double: u8 (0/1)
CMD_DOWN       = 0x03   # button_id: u8
CMD_UP         = 0x04   # button_id: u8
CMD_SCROLL     = 0x05   # dy: i16, dx: i16
CMD_ZOOM       = 0x06   # dir: i8
CMD_GESTURE    = 0x07   # gesture_id: u8
CMD_TEXT       = 0x08   # len: u16, utf8_bytes:[len]
CMD_KEY        = 0x09   # key_id: u8, mod_count: u8, mod_ids: [u8]
CMD_MEDIA      = 0x0A   # action_id: u8
CMD_VOLSET     = 0x0B   # volume: u8
CMD_VOLGET     = 0x0C   # (empty)
CMD_RADIO      = 0x0D   # device_id: u8
CMD_POWER      = 0x0E   # action_id: u8
CMD_BRIGHT     = 0x0F   # delta: i8
CMD_PING       = 0x10   # (empty)
CMD_SCROLLINFO = 0x13   # server -> client: posisi scroll window aktif

# ── Lookup tables ────────────────────────────────────────────────────────────
_BUTTON_IDS = {"left": 0, "right": 1, "middle": 2}
_BUTTON_NAMES = {v: k for k, v in _BUTTON_IDS.items()}

_GESTURE_IDS = {"taskview": 0, "showdesktop": 1, "appnext": 2, "appprev": 3}
_GESTURE_NAMES = {v: k for k, v in _GESTURE_IDS.items()}

_MEDIA_IDS = {
    "playpause": 0, "next": 1, "prev": 2, "stop": 3,
    "volup": 4, "voldown": 5, "mute": 6,
}
_MEDIA_NAMES = {v: k for k, v in _MEDIA_IDS.items()}

_RADIO_IDS = {"wifi": 0, "bluetooth": 1, "hotspot": 2}
_RADIO_NAMES = {v: k for k, v in _RADIO_IDS.items()}

_POWER_IDS = {"shutdown": 0, "restart": 1, "sleep": 2, "lock": 3}
_POWER_NAMES = {v: k for k, v in _POWER_IDS.items()}

# VK codes yang umum dipakai (sama dengan input_core.VK)
_VK_IDS = {
    "enter": 0x0D, "esc": 0x1B, "tab": 0x09, "backspace": 0x08, "delete": 0x2E,
    "space": 0x20, "up": 0x26, "down": 0x28, "left": 0x25, "right": 0x27,
    "home": 0x24, "end": 0x23, "pgup": 0x21, "pgdn": 0x22, "win": 0x5B,
    "ctrl": 0x11, "alt": 0x12, "shift": 0x10, "insert": 0x2D, "capslock": 0x14,
    "printscreen": 0x2C, "d": 0x44,
    **{f"f{i}": 0x6F + i for i in range(1, 13)},
}
_VK_NAMES = {v: k for k, v in _VK_IDS.items()}

# Key ID 0x00 = custom/UTF8, diikuti len:u16 + bytes
KEY_CUSTOM = 0x00


# ── Encoder ──────────────────────────────────────────────────────────────────

def encode(msg: dict) -> bytes | None:
    """
    Encode satu pesan protokol ke binary.  Kembalikan None kalau command
    tidak dikenal atau payload tidak valid — caller harus fallback ke JSON.
    """
    t = msg.get("t")

    if t == "move":
        dx = _clamp_i16(msg.get("dx", 0))
        dy = _clamp_i16(msg.get("dy", 0))
        return struct.pack("<Bhh", CMD_MOVE, dx, dy)

    if t == "click":
        bid = _BUTTON_IDS.get(msg.get("b", "left"), 0)
        dbl = 1 if msg.get("double", False) else 0
        return struct.pack("<BBB", CMD_CLICK, bid, dbl)

    if t == "down":
        bid = _BUTTON_IDS.get(msg.get("b", "left"), 0)
        return struct.pack("<BB", CMD_DOWN, bid)

    if t == "up":
        bid = _BUTTON_IDS.get(msg.get("b", "left"), 0)
        return struct.pack("<BB", CMD_UP, bid)

    if t == "scroll":
        dy = _clamp_i16(msg.get("dy", 0))
        dx = _clamp_i16(msg.get("dx", 0))
        return struct.pack("<Bhh", CMD_SCROLL, dy, dx)

    if t == "zoom":
        d = _clamp_i8(msg.get("dir", 1))
        return struct.pack("<Bb", CMD_ZOOM, d)

    if t == "gesture":
        gid = _GESTURE_IDS.get(msg.get("g", ""), None)
        if gid is None:
            return None
        return struct.pack("<BB", CMD_GESTURE, gid)

    if t == "text":
        text = msg.get("s", "")
        encoded = text.encode("utf-8")
        if len(encoded) > 65535:
            return None
        return struct.pack("<BH", CMD_TEXT, len(encoded)) + encoded

    if t == "key":
        k = msg.get("k", "")
        mods = msg.get("mods") or []
        vk = _VK_IDS.get(k.lower(), None)
        if vk is not None:
            mod_bytes = bytes([_VK_IDS.get(m.lower(), 0x11) for m in mods])
            if len(mod_bytes) > 255:
                return None
            return struct.pack("<BB", CMD_KEY, vk) + struct.pack("B", len(mod_bytes)) + mod_bytes
        # Custom key (UTF-8 character)
        key_bytes = k.encode("utf-8")
        if len(key_bytes) > 65535:
            return None
        mod_bytes = bytes([_VK_IDS.get(m.lower(), 0x11) for m in mods])
        if len(mod_bytes) > 255:
            return None
        header = struct.pack("<BBH", CMD_KEY, KEY_CUSTOM, len(key_bytes))
        return header + key_bytes + struct.pack("B", len(mod_bytes)) + mod_bytes

    if t == "media":
        mid = _MEDIA_IDS.get(msg.get("a", ""), None)
        if mid is None:
            return None
        return struct.pack("<BB", CMD_MEDIA, mid)

    if t == "volset":
        v = max(0, min(100, int(msg.get("v", 50))))
        return struct.pack("<BB", CMD_VOLSET, v)

    if t == "volget":
        return struct.pack("<B", CMD_VOLGET)

    if t == "radio":
        did = _RADIO_IDS.get(msg.get("d", ""), None)
        if did is None:
            return None
        return struct.pack("<BB", CMD_RADIO, did)

    if t == "power":
        aid = _POWER_IDS.get(msg.get("a", ""), None)
        if aid is None:
            return None
        return struct.pack("<BB", CMD_POWER, aid)

    if t == "bright":
        d = _clamp_i8(msg.get("d", 10))
        return struct.pack("<Bb", CMD_BRIGHT, d)

    if t == "ping":
        return struct.pack("<B", CMD_PING)

    return None


# ── Decoder ──────────────────────────────────────────────────────────────────

def decode(data: bytes) -> dict | None:
    """
    Decode satu paket binary ke dict JSON.  Kembalikan None kalau data
    tidak valid atau command ID tidak dikenal.
    """
    if len(data) < 1:
        return None

    cmd = data[0]

    try:
        if cmd == CMD_MOVE:
            _, dx, dy = struct.unpack_from("<Bhh", data, 0)
            return {"t": "move", "dx": dx, "dy": dy}

        if cmd == CMD_CLICK:
            _, bid, dbl = struct.unpack_from("<BBB", data, 0)
            return {"t": "click", "b": _BUTTON_NAMES.get(bid, "left"), "double": dbl == 1}

        if cmd == CMD_DOWN:
            _, bid = struct.unpack_from("<BB", data, 0)
            return {"t": "down", "b": _BUTTON_NAMES.get(bid, "left")}

        if cmd == CMD_UP:
            _, bid = struct.unpack_from("<BB", data, 0)
            return {"t": "up", "b": _BUTTON_NAMES.get(bid, "left")}

        if cmd == CMD_SCROLL:
            _, dy, dx = struct.unpack_from("<Bhh", data, 0)
            return {"t": "scroll", "dy": dy, "dx": dx}

        if cmd == CMD_ZOOM:
            _, d = struct.unpack_from("<Bb", data, 0)
            return {"t": "zoom", "dir": d}

        if cmd == CMD_GESTURE:
            _, gid = struct.unpack_from("<BB", data, 0)
            return {"t": "gesture", "g": _GESTURE_NAMES.get(gid, "")}

        if cmd == CMD_TEXT:
            _, tlen = struct.unpack_from("<BH", data, 0)
            text = data[3:3 + tlen].decode("utf-8")
            return {"t": "text", "s": text}

        if cmd == CMD_KEY:
            _, vk_or_zero = struct.unpack_from("<BB", data, 0)
            pos = 2
            if vk_or_zero == KEY_CUSTOM:
                klen = struct.unpack_from("<H", data, pos)[0]
                pos += 2
                k = data[pos:pos + klen].decode("utf-8")
                pos += klen
            else:
                k = _VK_NAMES.get(vk_or_zero, "")
                pos = 2
            mcount = data[pos]
            pos += 1
            mods = [_VK_NAMES.get(data[pos + i], "") for i in range(mcount)]
            return {"t": "key", "k": k, "mods": mods}

        if cmd == CMD_MEDIA:
            _, mid = struct.unpack_from("<BB", data, 0)
            return {"t": "media", "a": _MEDIA_NAMES.get(mid, "")}

        if cmd == CMD_VOLSET:
            _, v = struct.unpack_from("<BB", data, 0)
            return {"t": "volset", "v": v}

        if cmd == CMD_VOLGET:
            return {"t": "volget"}

        if cmd == CMD_RADIO:
            _, did = struct.unpack_from("<BB", data, 0)
            return {"t": "radio", "d": _RADIO_NAMES.get(did, "")}

        if cmd == CMD_POWER:
            _, aid = struct.unpack_from("<BB", data, 0)
            return {"t": "power", "a": _POWER_NAMES.get(aid, "")}

        if cmd == CMD_BRIGHT:
            _, d = struct.unpack_from("<Bb", data, 0)
            return {"t": "bright", "d": d}

        if cmd == CMD_PING:
            return {"t": "ping"}

    except (struct.error, IndexError, UnicodeDecodeError):
        return None

    return None


# ── Helpers ──────────────────────────────────────────────────────────────────

def _clamp_i16(v):
    return max(-32768, min(32767, int(v)))


def _clamp_i8(v):
    return max(-128, min(127, int(v)))

File: android/app/src/main/java/com/bjorn/claudepad/BinaryProtocol.kt
package com.bjorn.claudepad

import org.json.JSONArray
import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Binary protocol encoder/decoder — pasangan Kotlin dari server/binary_protocol.py.
 *
 * v3.3: Menggantikan JSON untuk command yang sering dikirim (move, scroll, click).
 * Paket 4-10x lebih kecil dari JSON, mengurangi latency & CPU.
 *
 * Wire format: [CMD_ID: 1 byte][PAYLOAD: 0..N bytes]
 * Semua integer multi-byte memakai little-endian.
 */
object BinaryProtocol {

    // ── Command IDs ──
    private const val CMD_MOVE     = 0x01
    private const val CMD_CLICK    = 0x02
    private const val CMD_DOWN     = 0x03
    private const val CMD_UP       = 0x04
    private const val CMD_SCROLL   = 0x05
    private const val CMD_ZOOM     = 0x06
    private const val CMD_GESTURE  = 0x07
    private const val CMD_TEXT     = 0x08
    private const val CMD_KEY      = 0x09
    private const val CMD_MEDIA    = 0x0A
    private const val CMD_VOLSET   = 0x0B
    private const val CMD_VOLGET   = 0x0C
    private const val CMD_RADIO    = 0x0D
    private const val CMD_POWER    = 0x0E
    private const val CMD_BRIGHT   = 0x0F
    private const val CMD_PING     = 0x10
    const val CMD_SCROLLINFO = 0x13  // server -> client: posisi scroll window aktif
    private const val KEY_CUSTOM   = 0x00

    // ── Lookup tables ──
    private val BUTTON_IDS = mapOf("left" to 0, "right" to 1, "middle" to 2)
    private val BUTTON_NAMES = BUTTON_IDS.entries.associate { (k, v) -> v to k }

    private val GESTURE_IDS = mapOf("taskview" to 0, "showdesktop" to 1, "appnext" to 2, "appprev" to 3)
    private val GESTURE_NAMES = GESTURE_IDS.entries.associate { (k, v) -> v to k }

    private val MEDIA_IDS = mapOf(
        "playpause" to 0, "next" to 1, "prev" to 2, "stop" to 3,
        "volup" to 4, "voldown" to 5, "mute" to 6
    )
    private val MEDIA_NAMES = MEDIA_IDS.entries.associate { (k, v) -> v to k }

    private val RADIO_IDS = mapOf("wifi" to 0, "bluetooth" to 1, "hotspot" to 2)
    private val RADIO_NAMES = RADIO_IDS.entries.associate { (k, v) -> v to k }

    private val POWER_IDS = mapOf("shutdown" to 0, "restart" to 1, "sleep" to 2, "lock" to 3)
    private val POWER_NAMES = POWER_IDS.entries.associate { (k, v) -> v to k }

    private val VK_IDS = mutableMapOf(
        "enter" to 0x0D, "esc" to 0x1B, "tab" to 0x09, "backspace" to 0x08,
        "delete" to 0x2E, "space" to 0x20, "up" to 0x26, "down" to 0x28,
        "left" to 0x25, "right" to 0x27, "home" to 0x24, "end" to 0x23,
        "pgup" to 0x21, "pgdn" to 0x22, "win" to 0x5B, "ctrl" to 0x11,
        "alt" to 0x12, "shift" to 0x10, "insert" to 0x2D, "capslock" to 0x14,
        "printscreen" to 0x2C, "d" to 0x44
    )
    private val VK_NAMES = mutableMapOf<Int, String>()

    init {
        // Isi VK_NAMES kebalikan dari VK_IDS + tambah F1..F13
        for ((k, v) in VK_IDS) VK_NAMES[v] = k
        for (i in 1..13) {
            VK_IDS["f$i"] = 0x6F + i
            VK_NAMES[0x6F + i] = "f$i"
        }
    }

    /** Encode JSON command ke binary. Kembalikan null kalau tidak bisa di-encode. */
    fun encode(msg: JSONObject): ByteArray? {
        val t = msg.optString("t")
        return when (t) {
            "move" -> {
                val dx = msg.optInt("dx", 0).coerceIn(-32768, 32767).toShort()
                val dy = msg.optInt("dy", 0).coerceIn(-32768, 32767).toShort()
                ByteBuffer.allocate(5).order(ByteOrder.LITTLE_ENDIAN)
                    .put(CMD_MOVE.toByte()).putShort(dx).putShort(dy).array()
            }
            "click" -> {
                val bid = BUTTON_IDS[msg.optString("b", "left")] ?: 0
                val dbl = if (msg.optBoolean("double", false)) 1 else 0
                byteArrayOf(CMD_CLICK.toByte(), bid.toByte(), dbl.toByte())
            }
            "down" -> {
                val bid = BUTTON_IDS[msg.optString("b", "left")] ?: 0
                byteArrayOf(CMD_DOWN.toByte(), bid.toByte())
            }
            "up" -> {
                val bid = BUTTON_IDS[msg.optString("b", "left")] ?: 0
                byteArrayOf(CMD_UP.toByte(), bid.toByte())
            }
            "scroll" -> {
                val dy = msg.optInt("dy", 0).coerceIn(-32768, 32767).toShort()
                val dx = msg.optInt("dx", 0).coerceIn(-32768, 32767).toShort()
                ByteBuffer.allocate(5).order(ByteOrder.LITTLE_ENDIAN)
                    .put(CMD_SCROLL.toByte()).putShort(dy).putShort(dx).array()
            }
            "zoom" -> {
                val d = msg.optInt("dir", 1).coerceIn(-128, 127).toByte()
                byteArrayOf(CMD_ZOOM.toByte(), d)
            }
            "gesture" -> {
                val gid = GESTURE_IDS[msg.optString("g", "")] ?: return null
                byteArrayOf(CMD_GESTURE.toByte(), gid.toByte())
            }
            "text" -> {
                val s = msg.optString("s", "")
                val encoded = s.toByteArray(Charsets.UTF_8)
                if (encoded.size > 65535) return null
                val buf = ByteBuffer.allocate(3 + encoded.size).order(ByteOrder.LITTLE_ENDIAN)
                buf.put(CMD_TEXT.toByte())
                buf.putShort(encoded.size.toShort())
                buf.put(encoded)
                buf.array()
            }
            "key" -> {
                val k = msg.optString("k", "")
                val mods = msg.optJSONArray("mods")
                val modList = mutableListOf<Byte>()
                if (mods != null) {
                    for (i in 0 until mods.length()) {
                        val m = mods.optString(i, "")
                        modList.add((VK_IDS[m.lowercase()] ?: 0x11).toByte())
                    }
                }
                if (modList.size > 255) return null
                val vk = VK_IDS[k.lowercase()]
                if (vk != null) {
                    val buf = ByteBuffer.allocate(3 + modList.size).order(ByteOrder.LITTLE_ENDIAN)
                    buf.put(CMD_KEY.toByte())
                    buf.put(vk.toByte())
                    buf.put(modList.size.toByte())
                    for (m in modList) buf.put(m)
                    return buf.array()
                }
                // Custom UTF-8 key
                val keyBytes = k.toByteArray(Charsets.UTF_8)
                if (keyBytes.size > 65535) return null
                val buf = ByteBuffer.allocate(5 + keyBytes.size + modList.size).order(ByteOrder.LITTLE_ENDIAN)
                buf.put(CMD_KEY.toByte())
                buf.put(KEY_CUSTOM.toByte())
                buf.putShort(keyBytes.size.toShort())
                buf.put(keyBytes)
                buf.put(modList.size.toByte())
                for (m in modList) buf.put(m)
                buf.array()
            }
            "media" -> {
                val mid = MEDIA_IDS[msg.optString("a", "")] ?: return null
                byteArrayOf(CMD_MEDIA.toByte(), mid.toByte())
            }
            "volset" -> {
                val v = msg.optInt("v", 50).coerceIn(0, 100).toByte()
                byteArrayOf(CMD_VOLSET.toByte(), v)
            }
            "volget" -> byteArrayOf(CMD_VOLGET.toByte())
            "radio" -> {
                val did = RADIO_IDS[msg.optString("d", "")] ?: return null
                byteArrayOf(CMD_RADIO.toByte(), did.toByte())
            }
            "power" -> {
                val aid = POWER_IDS[msg.optString("a", "")] ?: return null
                byteArrayOf(CMD_POWER.toByte(), aid.toByte())
            }
            "bright" -> {
                val d = msg.optInt("d", 10).coerceIn(-128, 127).toByte()
                byteArrayOf(CMD_BRIGHT.toByte(), d)
            }
            "ping" -> byteArrayOf(CMD_PING.toByte())
            else -> null
        }
    }
}

/** Informasi posisi scroll dari server (CMD_SCROLLINFO). */
data class ScrollInfo(val pos: Int, val min: Int, val max: Int, val page: Int)

/** Decode binary CMD_SCROLLINFO payload (16 bytes: 4 x little-endian Int). */
fun decodeScrollInfo(data: ByteBuffer): ScrollInfo? {
    if (data.remaining() < 16) return null
    return try {
        ScrollInfo(
            pos = data.getInt(),
            min = data.getInt(),
            max = data.getInt(),
            page = data.getInt()
        )
    } catch (e: Exception) { null }
}
